nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -16,88 +16,210 @@
+          master_array[index] = interpretValue(costmap_update_msg_->data[di++]);
+        }
+      }
+      has_update_ = false;
+    } else {
+      if (costmap_ == nullptr) {
+        costmap_ = std::make_shared<Costmap2D>(
+          costmap_msg_->info.width, costmap_msg_->info.height,
+          costmap_msg_->info.resolution, costmap_msg_->info.origin.position.x,
+          costmap_msg_->info.origin.position.y);
+      } else if (costmap_->getSizeInCellsX() != costmap_msg_->info.width ||  // NOLINT
+        costmap_->getSizeInCellsY() != costmap_msg_->info.height ||
+        costmap_->getResolution() != costmap_msg_->info.resolution ||
+        costmap_->getOriginX() != costmap_msg_->info.origin.position.x ||
+        costmap_->getOriginY() != costmap_msg_->info.origin.position.y)
+      {
+        // Update the size of the costmap
+        costmap_->resizeMap(costmap_msg_->info.width, costmap_msg_->info.height,
+          costmap_msg_->info.resolution,
+          costmap_msg_->info.origin.position.x,
+          costmap_msg_->info.origin.position.y);
+      }
 
-  unsigned char * master_array = costmap_->getCharMap();
-  unsigned int index = 0;
-  for (unsigned int i = 0; i < costmap_msg_->metadata.size_x; ++i) {
-    for (unsigned int j = 0; j < costmap_msg_->metadata.size_y; ++j) {
-      master_array[index] = costmap_msg_->data[index];
-      ++index;
+      unsigned char * master_array = costmap_->getCharMap();
+      unsigned int index = 0;
+      for (unsigned int i = 0; i < costmap_msg_->info.width; ++i) {
+        for (unsigned int j = 0; j < costmap_msg_->info.height; ++j) {
+          unsigned char value = costmap_msg_->data[index];
+          master_array[index] = interpretValue(value);
+          ++index;
+        }
+      }
     }
   }
 }
 
-void CostmapSubscriber::costmapCallback(const nav2_msgs::msg::Costmap::SharedPtr msg)
+unsigned char
+CostmapSubscriber::interpretValue(unsigned char value)
+{
+  // check if the static value is above the unknown or lethal thresholds
+  if (track_unknown_space_ && value == unknown_cost_value_) {
+    return NO_INFORMATION;
+  } else if (!track_unknown_space_ && value == unknown_cost_value_) {
+    return FREE_SPACE;
+  } else if (value >= lethal_threshold_) {
+    return LETHAL_OBSTACLE;
+  } else if (trinary_costmap_) {
+    return FREE_SPACE;
+  }
+
+  double scale = static_cast<double>(value) / lethal_threshold_;
+  return scale * LETHAL_OBSTACLE;
+}
+
+void CostmapSubscriber::costmapRawCallback(const nav2_msgs::msg::Costmap::SharedPtr msg)
+{
+  costmap_raw_msg_ = msg;
+  if (!costmap_received_) {
+    costmap_received_ = true;
+  }
+  toCostmap2D();
+}
+
+void CostmapSubscriber::costmapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
 {
   costmap_msg_ = msg;
   if (!costmap_received_) {
     costmap_received_ = true;
   }
+  toCostmap2D();
+}
+
+void CostmapSubscriber::costmapUpdateCallback(
+  const map_msgs::msg::OccupancyGridUpdate::SharedPtr msg)
+{
```

Comment:
You also need to check here that the timestamp of `msg` is **newer** than the last full costmap update. This is because if an old update is applied to a new costmap, it'll mess up a section of it.